[nja99] iP

nja99 opened the pull request:
# 🖍️Crayon Chatbot🖍️
> “If you spend too much time thinking about a thing, you’ll never get it done.” – Bruce Lee ([source](https://dansilvestre.com/productivity-quotes/#3_%E2%80%9CIf_you_spend_too_much_time_thinking_about_a_thing_youll_never_get_it_done%E2%80%9D_%E2%80%93_Bruce_Lee))

Meet **Crayon** a smart and intuitive chatbot designed to help you manage tasks effortlessly. Built for simplicity and efficiency, it's your personal assistant for a more productive day!

- CLI-Based
- Simple Commands
- ~~Fast~~ Lightning Fast

## Get Started
All you need to do is,

1. Download it from [here](https://github.com/nja99/ip)
2. Start by double-clicking the App.
3. Add your Tasks.
4. Let **Crayon** do the rest for you! 😉

## Features
- [x] Managing To Dos
- [x] Tracking Deadlines
- [x] Scheduling Events

## Developers
**Crayon** is a project built in _Java_ using **_Object-Oriented Programming (OOP)_** principles. 
Here is a snippet of the `main` method

``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

nja99 committed:
Renamed and implemeted initial skeletal for Crayon Chatbot

nja99 committed:
Customized greeting message, and added Echo support

nja99 committed:
Added ArrayList to keep track of tasks

nja99 committed:
Added support for list

nja99 committed:
Refactored task-related methods into separate TaskManager class

nja99 committed:
Refactored constants into Constants class

nja99 committed:
Added Task class for Task Management. Added ability to change task status

nja99 committed:
Added support for ToDo, Event and Deadline. Refactored Tasks

nja99 committed:
Added Automated Text UI Testing with input and expected output

nja99 committed:
Added Custom Exception Handling support for task management

nja99 committed:
Add support for deleting tasks from the list. Refactored validation of taskId

nja99 committed:
Minor code fixes

nja99 committed:
Refactored code and Added enums

nja99 committed:
Add CSVWriter.java

nja99 committed:
Remove unused import statement

nja99 committed:
Add CSVReader.java

Add: Functions to ToDo.java, Event.java, Deadline.java to create Task from CSV row

nja99 committed:
Move: CSVReader.java, CSVWriter.java to utils

nja99 committed:
Bug Fix: Modify data parsing in creating Task from CSV

nja99 committed:
Add DateTime.java

Modify: Event.java, Deadline.java to store startDate and endDate in LocalDateTime format.

nja99 committed:
Modify: input.txt and EXPECTED.TXT with new test cases

nja99 committed:
Refactor: TaskManager.java into TaskList.java

nja99 committed:
Add CrayonException.java

Modify: CrayonIllegalArgumentException.java, CrayonIllegalArgumentException.java, CrayonUnsupportedTaskException.java to extend from CrayonException.java

nja99 committed:
Add Storage.java

Modify: CSVWriter.java and CSVReader.java to make it more generic

nja99 committed:
Move: All files into java/crayon

nja99 committed:
Add Commands to handle user actions

nja99 committed:
Add Ui.java

Refactor printTaskAction and printStatusAction in TaskList.java into Ui.java

nja99 committed:
Add Parser.java

Refactored echoMessage from Crayon.java into Parser.java to handle user inputs

nja99 committed:
Minor Bug Fix: Remove wildcard imports

nja99 committed:
Changes: Add AutoClosable to Ui to ensure Ui closes on exit.

nja99 committed:
Minor Fixes: Fix error message format

nja99 committed:
Merge branch 'add-gradle-support'
Merge Gradle Support

nja99 committed:
Configure gradle build automation

nja99 committed:
Bug Fix: Fix all checkstyle error

nja99 committed:
Add: JavaDocs for package commands

nja99 committed:
Add: JavaDocs for package exceptions

nja99 committed:
Add: JavaDocs for packages enums and utils

nja99 committed:
Add: JavaDocs for package tasks

nja99 committed:
Add: JavaDocs for multiple files

Add JavaDocs for Constants.java, Crayon.java, Parser.java, Storage.java, TaskList.java and Ui.java

nja99 committed:
DateTime.java: Fix JavaDoc for function

nja99 committed:
Resolve Merge Conflicts with branch-A-CodingStandard

nja99 committed:
Add FindCommand.java

Add filterTasks and listFilteredTasks in TaskList.java

nja99 committed:
Merge branch-Level-9 into master

nja99 committed:
Fix checkstyle violations

nja99 committed:
Merge branch 'branch-Checkstyle'
Merge checkstyle fixes into master

nja99 committed:
Fix checkstyle.xml for lambda-style switch statements

nja99 committed:
Merge branch 'branch-Checkstyle'

monobeartae commented on src/main/java/crayon/Storage.java:
```diff
@@ -0,0 +1,71 @@
+            TaskType taskType = TaskType.fromString(values[0]);
+            try {
+                Task task = switch(taskType) {
+                case TODO -> createToDoFromCsv(values);
```
there can be a indentation in the switch case here, maybe to be consistent with your other switch cases

monobeartae reviewed (commented):
LGTM! :D

monobeartae commented on src/main/java/crayon/Constants.java:
```diff
@@ -0,0 +1,18 @@
+          ____ ____     _ __   _____  _   _
+         / ___|  _ \\   / \\\\ \\ / / _ \\| \\ | |
+        | |   | |_) | / _ \\\\ V | | | |  \\| |
+        | |___|  _ < / ___ \\| || |_| | |\\  |
```
omg you actually made the ascii for your bot name!! effort, looks great!

monobeartae commented on src/main/java/crayon/TaskList.java:
```diff
@@ -0,0 +1,178 @@
+    private static final String TASK_ADDED_MESSAGE = "Got it. I've added this task";
+    private static final String TASK_REMOVED_MESSAGE = "Noted. I've removed this task";
+    private static final String TASK_DONE_MESSAGE = "Nice! I've marked this task as done";
+    private static final String TASK_UNDONE_MESSAGE = "OK, I've marked this task as not done yet";
```
I like that you abstracted the messages up here to be easier to edit and more organised! really neat

monobeartae reviewed (commented):
wanted to point out things I liked too!

nja99 committed:
Package Storage.java, Parser.java, TaskList.java, Ui.java

nja99 committed:
Add StorageTest.java

Add JUnit test for Storage

nja99 committed:
Add ParserTest.java

nja99 committed:
Add JavaDocs for JUnit Test

Add JavaDocs for JUnit test and fix checkstyle

nja99 committed:
Merge branch 'branch-A-JUnit'
Merge branch with "branch-A-JUnit"

Mingyuan03 commented on src/main/java/crayon/commands/Command.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    public boolean exitRequested() {
+        return this.isExit;
+    }
```
Just some nits to comment: perhaps the method name might sound more intuitive if it is `getExitStatus` since this is an accessor.

Mingyuan03 commented on src/test/java/crayon/storage/TaskStub.java:
Nice creation of stubs for JUnit!

Mingyuan03 commented on src/main/java/crayon/tasks/Task.java:
```diff
@@ -0,0 +1,65 @@
+ * Represents a task.
+ */
+public abstract class Task {
+    protected String description;
```
`description` field can be declared `final`.

Mingyuan03 commented on src/main/java/crayon/utils/DateTime.java:
```diff
@@ -0,0 +1,69 @@
+     */
+    public static LocalDateTime parseStoredDateTime(String storedDateTime) throws CrayonInvalidDateTimeException {
+        try {
+            return LocalDateTime.parse(storedDateTime);
```
Just a design opinion: `LocalDateTime::parse` would expect the datetime format to be strictly `ISO_LOCAL_DATE_TIME`, failing which the catch block is entered. Perhaps the implementation can lend itself to greater flexibility with `LocalDateTime.parse(storedDateTime).truncatedTo(ChronoUnit.SECONDS)`.

Mingyuan03 reviewed (commented):
LGTM. Good application of Java coding standards and detailed documentation makes your overall code a great pleasure to review. Besides some minor suggestions, I have confidence that your Crayon bot will be an impressive end-product! 😄

nja99 committed:
Add JavaFX to replace text-based UI

nja99 committed:
Refactor code removing unused functions

nja99 committed:
Remove Constants.java

nja99 committed:
Refactor Commands

nja99 committed:
Refactor: Move UI related components from TaskList.java to UI.java

nja99 committed:
Merge branch 'branch-Level-10'
Add JavaFX

nja99 committed:
Add CrayonTaskCreationException.java

nja99 committed:
Add JavaDoc for Crayon.java

nja99 committed:
Add assertion for multiple classes

Added assertion to ensure User Input is not Null or Empty before parsing.
Included assertion to verify the command execution returns a valid response
Added assertion to ensure that task list is not null before saving it to storage.

These assertions help to catch potential bugs during development and ensure that the program behaves as expected in different scenarios.

nja99 committed:
Add image resources for JavaFX

nja99 committed:
Merge branch 'branch-Level-10'
Added Images for JavaFX UI

nja99 committed:
Refactor Main.java method to adhere to SLAP

- Moved logic for setting up the main window into a dedicated method.
- Isolated application shutdown logic into a separate method
- Improved overall structure for easier extensibility and better understanding of the flow.

nja99 committed:
Refactor Parser.java to adhere to SLAP

- Split parsing of action and content into separate methods.
- Move command creation into its own method.
- Simplified parseCommand method by delegating tasks to helper methods.

nja99 committed:
Refactor Storage.java to adhere to SLAP

- Separated logic for preparing tasks rows and writing them to CSV.
- Separated logic for reading and creating task from CSV.
- Split task creation from CSV rows into dedicated method.

nja99 committed:
Refactor TaskList.java to adhere to SLAP

- Separated task creation logic into a dedicated method createTaskOfType.

nja99 committed:
Refactor DialogBox.java to adhere to SLAP

- Extracted FXML loading into its own method.
- Extracted setting of Dialog Content into its own method.

nja99 committed:
Merge pull request #1 from nja99/branch-A-Assertions

Add assertion for multiple classes

nja99 committed:
Merge branch 'master'

- Resolve merge conflicts in file Parser.java, Storage.java and TaskList.java
- Added getDoneStatus function in Task.java
- Remove an assertion in Storage.java

nja99 committed:
Merge pull request #2 from nja99/branch-A-CodeQuality

Improve Code Quality

nja99 committed:
Add gradle.yml for Github Actions Continuous Integration

nja99 committed:
Add Friendlier Syntax for commands

nja99 committed:
Implement application exit on ByeCommand

-Add functionality to close application when ByeCommand is executed.
-Display farewell message and wait 3 second before actual exit
-Save all tasks on exit

nja99 committed:
Enhance ListCommand to support task type filtering and sorting

- Implement task filtering by type (todo, deadline, event)
- Add sorting functionality for deadlines (by end date) and events (by start date)
This enhancement allows users to list tasks by specific types, providing a more intuitive and powerful task management experience.
The new syntax supports commands like "list todo", "list deadline", and "list event", making it easier for users to focus on specific task categories.

nja99 committed:
Merge pull request #4 from nja99/BCD-Extensions

BCD Extensions

nja99 committed:
Fix: sortTasks in TaskList.java

Issue where sortTasks was recursively calling itself.

nja99 committed:
Add: Extensive JUnit Tests

Added more JUnit Test into the 2 Test Class

nja99 committed:
Fix: Add JavaDocs and Fix CheckStyle

Added JavaDocs and Fixed CheckStyle for ParserTest.java, StorageTest.java, TaskStub.java

nja99 committed:
Add: App Icon and image sources

- Included attribution for image sources

nja99 committed:
Update: README.md

nja99 committed:
Update README.md for User Guide

nja99 committed:
Add JavaDocs for new functions

nja99 committed:
Fix: Style Add/Remove whitespaces

nja99 committed:
Bug Fix: TaskList.java fix TaskID range

nja99 committed:
Minor Fix: Fix Checkstyle in TaskType.java, Action.java

nja99 committed:
Add TaskType validation to ListCommand.java

- Implemented validation to ignore non-existent task types.
